### ISSUE TITLE
iOS: Refresh token is missing but it's impossible to sign out or clean the session

### ISSUE THREAD

**Issue by OlKir** [aws-amplify/amplify-flutter#2645]
### Description

On one of my testing devices local session become broken and can't get device out of this state. In simulator, on few other devices and on android integration with Cognito works as expected. I'm able to signup and signin using e-mail credentials or federated providers as Facebook or Apple ID.

But on one testing device any call to Amplify/Cognito services app shows this dialog:
<img src="https://user-images.githubusercontent.com/5834531/218399424-0fc73cc2-d634-47e5-9898-32b93fde2e93.jpg" width="300" />

`CognitoAuthSession` instance returned by `Amplify.Auth.fetchAuthSession` has `isSignedIn` set to true.
I can get userName - it's a Facebook user.
But when I debug iOS part of the code I see that in method 
```
- (void)getSessionInternal: (nullable UIViewController *) vc completion: (AWSCognitoAuthGetSessionBlock) completion
```
data returned from keychain are incomplete. After this assignment
```
        NSString * expirationDateKey = [self keyChainKey:keyChainNamespace key:AWSCognitoAuthUserTokenExpiration];
        NSString * expirationDate = self.keychain[expirationDateKey];
        NSString * scopesKey = [self keyChainKey:keyChainNamespace key:AWSCognitoAuthUserScopes];
        NSString * scopes = self.keychain[scopesKey];
```
`expirationDate` and `scopes` are nil. This causes presentation of general SignIn UI via `SFAuthenticationSession`.
Trying to use any sign in method on the page fails with error message "Safari cannot open the page because the address is invalid." So I can only close this dialog via Cancel that leads to "Error Domain=com.apple.SafariServices.Authentication Code=1"

The problem is that this dialog appears even when I try to use `Amplify.Auth.signOut()` and then sign out fails.
And even when I try to delete local user with `Amplify.Auth.deleteUser()` operation fails with the same error message.
If I try to use signIn methods framework complains that there is signed user and it should sign out first.
Reinstalling the app doesn't help as the session is stored in keychain.

**My preferred solution in the case when there is no refresh token would be just sign out local user and clean the session.**
I have to admit that I have no idea how device ended up in this state. It's a testing one so I signed in/out and installed/deleted the app many times using IDE and TestFlight.
I would be ready to accept it as a corner case but I need some way to get out of it and would be grateful if somebody give me a hint how else I can invalidate/clean local session.

 

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Storage

### Platforms

- [X] iOS
- [ ] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.7.3

### Amplify Flutter Version

0.6.12

### Deployment Method

Amplify CLI

**Comment by Jordan-Nelson** [comment]
@OlKir - I'm sorry that you are facing this issue. Keychain data is not removed when apps are uninstalled. This behavior is controlled by iOS/Apple.

On a simulator, you can go to Device -> clear all content and settings to wipe all data including keychain data. The options on a physical device are more limited. If this is a test device only, one option may be to factory reset it.

It sounds like you aren't sure how the device got into this state. If you have any steps that can reproduce this, or if this is a repeated issue, please let me know and we can try to reproduce it.

**Comment by OlKir** [comment]
@Jordan-Nelson Thank you for the response! 

I'm currently investigating on the physical device if it would be possible to clear data from keychain using framework code on iOS side.
Maybe I will find something useful this way, as currently inside Flutter `Amplify.Auth` still considers user to be signed in, when on iOS side framework doesn't recognize user after I executed `clearAll`. I will post my findings if I have any.

**Comment by OlKir** [comment]
@dnys1 Thank you for the support! I would like to try developer preview release but unfortunately it requires null safety and my app doesn't have it yet. But the version with Cognito-based functionality has not been published yet so I can experiment on it. I might put together separate small app using configuration from the main project to check.

**Comment by OlKir** [comment]
Sorry, I have not tested if developer preview release works as I managed to clean device keychain using plugin's code.
The system is pretty complex but I believe that some issue happens here:
<img width="888" alt="Screenshot 2023-02-18 at 14 53 32" src="https://user-images.githubusercontent.com/5834531/219870448-b3e75480-8ceb-4c71-922e-f1c607b537dc.png">
`currentUserState` set to `.signedIn` based on `self.internalCredentialsProvider?.identityId` although I would expect it to be also dependant on existence of refresh token.

And calling `.clearKeychain()` on `AWSMobileClient` doesn't help as user still considered signed in based on `cachedLoginsMap` which apparently remains in keychain.

Only after I commented loading of `cachedLoginsMap` out app behaved as expected: user state was signed out and I was able to sign in and from now on Cognito just works fine.

**Comment by Jordan-Nelson** [comment]
Hi @OlKir - I am glad you were able to get the test device working again. It sounds like your test device was in a state where the refresh token was null and the identityId was non-null. I believe this is not an expected state. If you find that the device ends up in this state again, let us know and we can try to determine what is causing that. As Dillon mentioned, the dev-preview version should be more resilient since an uninstall/re-install will clear the keychain.

Let me know if you have any other questions. Otherwise I will close this out.

**Comment by Jordan-Nelson** [comment]
@OlKir - I am going to close this out since I have  not heard back. If you face this issue again, please open a new issue. Thanks.
